File: packages/contracts/src/dataset/dataset.ts
/**
 * CCTV dataset + accuracy-evaluation contracts (AI-5e, Architect priorities 1 & 3). These describe
 * **recorded surveillance footage and what the platform is expected to see in it** — the evidence
 * that moves a capability from "runs" to "works", which no amount of simulation can supply.
 *
 * Governance shapes only: no runtime behavior depends on them and the five frozen perception
 * contracts are untouched. The runtime mirrors these in ai/inference/dataset.py + evaluation.py.
 *
 * **Footage bytes never live in git.** A `DatasetCase` is a manifest — it points at footage held in
 * DVC/object storage (ai/datasets/README.md) and records consent/licence for it, because surveillance
 * footage of real people is exactly the kind of data that must not be casually copied around.
 */
import { z } from 'zod';
import { IsoDateTime, SemVer } from '../common/primitives.js';
import { EvidenceClass } from '../certification/certification.js';

/**
 * The scenario a dataset case exercises. This is the Architect's AI-5e priority-1 list, and it is
 * deliberately a **closed enum**: an open string would let cases accumulate under seven spellings of
 * "loitering" and quietly stop being a regression suite. New scenarios are added additively here.
 */
export const ScenarioCategory = z.enum([
  'person_detection',
  'tracking',
  'queue',
  'crowd',
  'loitering',
  'intrusion',
  'restricted_area',
  'shoplifting',
  'cashier_theft',
  'suspicious_behavior',
  'fire',
  'smoke',
  'violence',
  'abandoned_object',
  'fall_detection',
  'ppe',
  'customer_movement',
  'staff_movement',
]);
export type ScenarioCategory = z.infer<typeof ScenarioCategory>;

/**
 * Where the footage actually is, and whether we are allowed to have it. `path` is repo-relative and
 * resolves to a DVC-tracked file; `sha256` pins the exact bytes so an evaluation result can never be
 * attributed to footage that has since been replaced.
 */
export const FootageReference = z.object({
  /** Repo-relative path under ai/datasets/. The bytes are DVC-tracked, not committed. */
  path: z.string().min(1).max(500),
  /** Digest of the footage. When present it is verified before evaluation — results pin to bytes. */
  sha256: z
    .string()
    .regex(/^[0-9a-f]{64}$/, 'must be a lowercase hex sha256')
    .optional(),
  durationSeconds: z.number().positive().optional(),
  fps: z.number().positive().optional(),
  resolution: z
    .string()
    .regex(/^\d{2,5}x\d{2,5}$/, 'must be WIDTHxHEIGHT')
    .optional(),
  /** Where it came from: `public-dataset`, `synthetic`, `customer-pilot`, `internal-capture`. */
  origin: z.enum(['public-dataset', 'synthetic', 'customer-pilot', 'internal-capture']),
  /** Licence or consent basis. Required — footage of real people without a recorded basis is not usable. */
  licence: z.string().min(1).max(200),
  /** Anonymisation applied (face blur, plate blur, …), when any. */
  anonymisation: z.string().max(500).optional(),
});
export type FootageReference = z.infer<typeof FootageReference>;

/**
 * One thing the platform is expected to produce from the footage, with the window it is expected in.
 * Expectations are **temporal and countable, not pixel-exact**: the platform's job is to notice that
 * someone loitered near the counter between 0:12 and 0:31, and holding it to a bounding-box IoU it was
 * never labelled for would make the suite expensive to maintain and dishonest about what it proves.
 */
export const ExpectedOccurrence = z.object({
  /** What kind of output this expects. */
  kind: z.enum(['detection', 'track', 'behavior', 'composite', 'event', 'incident']),
  /** Label / behaviorType / eventType, e.g. `person`, `loitering`, `security.intrusion.detected`. */
  type: z.string().min(1).max(120),
  /** Expected count, when the case is about how many. */
  count: z.number().int().nonnegative().optional(),
  /** Inclusive tolerance on `count` — real footage does not produce exact numbers. */
  countTolerance: z.number().int().nonnegative().default(0),
  /** Window (seconds from footage start) the occurrence is expected within. */
  fromSeconds: z.number().nonnegative().optional(),
  toSeconds: z.number().nonnegative().optional(),
  /** Zone the occurrence is expected in, when the case defines zones. */
  zoneId: z.string().max(120).optional(),
  /** Minimum confidence the platform should report. */
  minConfidence: z.number().min(0).max(1).optional(),
  /**
   * A **negative** expectation: this must NOT occur. False positives are the failure mode that
   * loses customer trust in surveillance analytics, so the suite has to be able to assert their
   * absence as directly as it asserts a presence.
   */
  absent: z.boolean().default(false),
  notes: z.string().max(500).optional(),
});
export type ExpectedOccurrence = z.infer<typeof ExpectedOccurrence>;

/**
 * A single evaluable piece of footage plus what it should produce. This is the unit of the CCTV
 * dataset library and the unit of accuracy regression.
 */
export const DatasetCase = z.object({
  /** Stable slug, e.g. `loitering/entrance-dwell-01`. */
  id: z.string().min(1).max(200),
  category: ScenarioCategory,
  title: z.string().min(1).max(200),
  description: z.string().max(2000).optional(),
  footage: FootageReference,
  /** Zones the case defines, in the generic zone-config shape the analyzer already consumes. */
  zones: z.array(z.record(z.string(), z.unknown())).default([]),
  /** BehaviorProfile name to run the case under, when it needs one. */
  profile: z.string().max(120).optional(),
  /** Analyzer options for the run (confidence, tracking, behavior thresholds). */
  options: z.record(z.string(), z.unknown()).default({}),
  expectations: z.array(ExpectedOccurrence).default([]),
  /** Free-form notes: what makes this clip hard, what a human sees in it. */
  notes: z.string().max(2000).optional(),
  /** Known gaps — recorded so the suite documents its own limits instead of hiding them. */
  futureImprovements: z.array(z.string().max(300)).default([]),
  addedAt: IsoDateTime.optional(),
});
export type DatasetCase = z.infer<typeof DatasetCase>;

/** How one expectation actually turned out. */
export const EvaluationOutcome = z.enum([
  /** Produced as expected, within tolerance. */
  'match',
  /** Expected but never produced — a false negative. */
  'missing',
  /** Asserted absent but produced anyway — a false positive. */
  'unexpected',
  /** Produced, but the count/timing/confidence fell outside tolerance. */
  'out-of-tolerance',
  /**
   * Recorded but not judged by this evaluator. A `DatasetCase` describes what the WHOLE PLATFORM
   * should make of a clip, including the incident a rule would raise — but the AI runtime emits
   * events, not incidents, and scoring itself against an output it does not own would be meaningless
   * in both directions. Deferred findings are reported and excluded from precision/recall.
   */
  'deferred',
]);
export type EvaluationOutcome = z.infer<typeof EvaluationOutcome>;

/** One expectation, judged. */
export const EvaluationFinding = z.object({
  kind: ExpectedOccurrence.shape.kind,
  type: z.string().min(1).max(120),
  outcome: EvaluationOutcome,
  expectedCount: z.number().int().nonnegative().optional(),
  observedCount: z.number().int().nonnegative().default(0),
  /** When timing mattered: the window asserted, and when it actually happened. */
  window: z.string().max(60).optional(),
  observedAtSeconds: z.number().nonnegative().optional(),
  detail: z.string().max(1000).optional(),
});
export type EvaluationFinding = z.infer<typeof EvaluationFinding>;

/**
 * Status of an evaluation run. **`footage-missing` is its own status and is never a pass** — the
 * dataset library must not turn into a suite that reports green because it evaluated nothing. It is
 * the expected status in CI, where the DVC bytes are not pulled.
 */
export const EvaluationStatus = z.enum(['pass', 'fail', 'footage-missing', 'error']);
export type EvaluationStatus = z.infer<typeof EvaluationStatus>;

/** The result of running one `DatasetCase` through the pipeline. */
export const EvaluationReport = z.object({
  id: z.string().min(1),
  caseId: z.string().min(1).max(200),
  category: ScenarioCategory,
  status: EvaluationStatus,
  runtimeVersion: SemVer,
  /** Model/engine the case ran under — accuracy claims are meaningless without it. */
  model: z.string().max(200).optional(),
  engine: z.string().max(60).optional(),
  findings: z.array(EvaluationFinding).default([]),
  truePositives: z.number().int().nonnegative().default(0),
  falseNegatives: z.number().int().nonnegative().default(0),
  falsePositives: z.number().int().nonnegative().default(0),
  /** tp / (tp + fp); null when nothing was produced or expected. */
  precision: z.number().min(0).max(1).nullable().default(null),
  /** tp / (tp + fn); null when nothing was expected. */
  recall: z.number().min(0).max(1).nullable().default(null),
  f1: z.number().min(0).max(1).nullable().default(null),
  framesProcessed: z.number().int().nonnegative().default(0),
  durationSeconds: z.number().nonnegative().default(0),
  /** `recorded-footage` for a real clip; `simulated` for a synthetic one. Never `hardware`. */
  evidenceClass: EvidenceClass,
  recordedAt: IsoDateTime,
  notes: z.string().max(2000).optional(),
});
export type EvaluationReport = z.infer<typeof EvaluationReport>;

/**
 * The aggregate across a dataset run — what a regression gate reads. `regressed` names the cases that
 * got worse against the accepted accuracy baseline, which is the only number that should block a merge.
 */
export const EvaluationSummary = z.object({
  id: z.string().min(1),
  runtimeVersion: SemVer,
  cases: z.number().int().nonnegative().default(0),
  passed: z.number().int().nonnegative().default(0),
  failed: z.number().int().nonnegative().default(0),
  /** Cases whose footage was not available locally — reported, never counted as passing. */
  skipped: z.number().int().nonnegative().default(0),
  truePositives: z.number().int().nonnegative().default(0),
  falseNegatives: z.number().int().nonnegative().default(0),
  falsePositives: z.number().int().nonnegative().default(0),
  precision: z.number().min(0).max(1).nullable().default(null),
  recall: z.number().min(0).max(1).nullable().default(null),
  f1: z.number().min(0).max(1).nullable().default(null),
  /** Per-category rollup, so a regression can be attributed to a scenario. */
  byCategory: z.record(z.string(), z.record(z.string(), z.number())).default({}),
  /** Case ids that scored worse than the accepted baseline. */
  regressed: z.array(z.string().min(1)).default([]),
  /** Case ids that scored better. */
  improved: z.array(z.string().min(1)).default([]),
  /** Overall gate: no failures and no regressions. Skipped cases do not make this true. */
  accepted: z.boolean().default(false),
  evidenceClass: EvidenceClass,
  recordedAt: IsoDateTime,
});
export type EvaluationSummary = z.infer<typeof EvaluationSummary>;

File: ai/inference/dataset.py
"""The CCTV dataset library (AI-5e, Architect priorities 1 & 3).

**What problem this solves.** The runtime has been proved against simulated sources and stub adapters.
That proves it *runs*. Whether it actually notices someone loitering by a till is a different question,
and the only thing that can answer it is real surveillance footage with a human's judgement written
down beside it. This module is the catalogue of those clips and those judgements.

**A case is a manifest, not a video.** Footage bytes live in DVC/object storage (see
ai/datasets/README.md) and are never committed — surveillance footage of real people is exactly the
category of data that must not be casually copied into a git history. A case therefore records where
the footage is, under what licence or consent basis, and what the platform should make of it. When the
bytes are not present locally the case reports `footage-missing`, which is **never** a pass: a dataset
suite that goes green because it evaluated nothing is worse than no suite at all.

**Why the category list is closed.** `SCENARIO_CATEGORIES` mirrors the contract enum. An open string
would let cases accumulate under three spellings of "loitering" and quietly stop functioning as a
regression suite — the failure mode of every test corpus that was ever allowed to grow organically.

Stdlib-only. Nothing here touches the perception pipeline; a dataset case is configuration + assertions.
"""

from __future__ import annotations

import hashlib
import json
import os
from dataclasses import dataclass, field
from typing import Dict, Iterable, Iterator, List, Optional, Sequence

#: Mirrors `ScenarioCategory` in @vip/contracts. Order is the order the library reports in.
SCENARIO_CATEGORIES = (
    "person_detection",
    "tracking",
    "queue",
    "crowd",
    "loitering",
    "intrusion",
    "restricted_area",
    "shoplifting",
    "cashier_theft",
    "suspicious_behavior",
    "fire",
    "smoke",
    "violence",
    "abandoned_object",
    "fall_detection",
    "ppe",
    "customer_movement",
    "staff_movement",
)

FOOTAGE_ORIGINS = ("public-dataset", "synthetic", "customer-pilot", "internal-capture")
EXPECTATION_KINDS = ("detection", "track", "behavior", "composite", "event", "incident")

#: Repo-relative root of the library. Cases live at `<root>/<category>/cases/<slug>.json`.
DATASETS_ROOT = os.path.abspath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "datasets")
)


class DatasetError(ValueError):
    """A malformed dataset manifest. Raised eagerly at load time: a case that cannot be parsed must
    not silently disappear from the suite, because a silently-absent regression test is the same as a
    passing one."""


@dataclass(frozen=True)
class FootageReference:
    """Where the footage is and whether we are allowed to have it."""

    path: str
    origin: str
    licence: str
    sha256: Optional[str] = None
    duration_seconds: Optional[float] = None
    fps: Optional[float] = None
    resolution: Optional[str] = None
    anonymisation: Optional[str] = None

    def __post_init__(self) -> None:
        if self.origin not in FOOTAGE_ORIGINS:
            raise DatasetError(f"unknown footage origin '{self.origin}'")
        if not self.licence.strip():
            raise DatasetError("footage must record a licence or consent basis")

    def resolve(self, root: Optional[str] = None) -> str:
        """Absolute path to the footage, whether or not it exists."""
        base = root or os.path.dirname(DATASETS_ROOT)
        return self.path if os.path.isabs(self.path) else os.path.abspath(os.path.join(base, self.path))

    def available(self, root: Optional[str] = None) -> bool:
        return os.path.isfile(self.resolve(root))

    def verify(self, root: Optional[str] = None) -> Optional[bool]:
        """Check the footage digest. `None` when the case declares none (nothing to verify) or the
        file is absent; `False` means the bytes changed and any result derived from them is void."""
        if not self.sha256 or not self.available(root):
            return None
        digest = hashlib.sha256()
        with open(self.resolve(root), "rb") as fh:
            for chunk in iter(lambda: fh.read(1024 * 1024), b""):
                digest.update(chunk)
        return digest.hexdigest() == self.sha256

    def to_dict(self) -> dict:
        out = {"path": self.path, "origin": self.origin, "licence": self.licence}
        for key, value in (
            ("sha256", self.sha256),
            ("durationSeconds", self.duration_seconds),
            ("fps", self.fps),
            ("resolution", self.resolution),
            ("anonymisation", self.anonymisation),
        ):
            if value is not None:
                out[key] = value
        return out


@dataclass(frozen=True)
class Expectation:
    """One thing the platform should — or must not — produce from the footage.

    Expectations are **temporal and countable, not pixel-exact**. The platform's job is to notice that
    someone dwelt near the counter between 0:12 and 0:31; holding it to a bounding-box IoU it was never
    labelled for would make the corpus expensive to maintain and dishonest about what it proves.
    """

    kind: str
    type: str
    count: Optional[int] = None
    count_tolerance: int = 0
    from_seconds: Optional[float] = None
    to_seconds: Optional[float] = None
    zone_id: Optional[str] = None
    min_confidence: Optional[float] = None
    absent: bool = False
    notes: Optional[str] = None

    def __post_init__(self) -> None:
        if self.kind not in EXPECTATION_KINDS:
            raise DatasetError(f"unknown expectation kind '{self.kind}'")
        if self.absent and self.count:
            raise DatasetError(
                f"expectation '{self.type}' is both absent and expects {self.count} occurrences"
            )
        if (
            self.from_seconds is not None
            and self.to_seconds is not None
            and self.to_seconds < self.from_seconds
        ):
            raise DatasetError(f"expectation '{self.type}' has an inverted window")

    @property
    def window(self) -> Optional[str]:
        if self.from_seconds is None and self.to_seconds is None:
            return None
        lo = "start" if self.from_seconds is None else f"{self.from_seconds:.1f}"
        hi = "end" if self.to_seconds is None else f"{self.to_seconds:.1f}"
        return f"{lo}-{hi}s"

    def within(self, seconds: Optional[float]) -> bool:
        """Whether an occurrence at `seconds` falls in the asserted window. An occurrence with no
        timestamp is accepted — a missing timestamp is the evaluator's gap, not the runtime's fault,
        and failing a case over it would produce noise instead of signal."""
        if seconds is None:
            return True
        if self.from_seconds is not None and seconds < self.from_seconds:
            return False
        if self.to_seconds is not None and seconds > self.to_seconds:
            return False
        return True

    def satisfied_by(self, observed: int) -> bool:
        if self.absent:
            return observed == 0
        target = 1 if self.count is None else self.count
        return abs(observed - target) <= self.count_tolerance

    def to_dict(self) -> dict:
        out: dict = {
            "kind": self.kind,
            "type": self.type,
            "countTolerance": self.count_tolerance,
            "absent": self.absent,
        }
        for key, value in (
            ("count", self.count),
            ("fromSeconds", self.from_seconds),
            ("toSeconds", self.to_seconds),
            ("zoneId", self.zone_id),
            ("minConfidence", self.min_confidence),
            ("notes", self.notes),
        ):
            if value is not None:
                out[key] = value
        return out


@dataclass
class DatasetCase:
    """One evaluable clip plus what it should produce."""

    id: str
    category: str
    title: str
    footage: FootageReference
    description: Optional[str] = None
    zones: List[dict] = field(default_factory=list)
    profile: Optional[str] = None
    options: dict = field(default_factory=dict)
    expectations: List[Expectation] = field(default_factory=list)
    notes: Optional[str] = None
    future_improvements: List[str] = field(default_factory=list)
    added_at: Optional[str] = None
    #: Where this case was loaded from — used by the CLI to point at the file that needs editing.
    manifest_path: Optional[str] = None

    def __post_init__(self) -> None:
        if self.category not in SCENARIO_CATEGORIES:
            raise DatasetError(
                f"case '{self.id}' declares unknown category '{self.category}'; "
                f"add it to SCENARIO_CATEGORIES and the ScenarioCategory contract first"
            )

    @property
    def positive_expectations(self) -> List[Expectation]:
        return [e for e in self.expectations if not e.absent]

    @property
    def negative_expectations(self) -> List[Expectation]:
        return [e for e in self.expectations if e.absent]

    def available(self, root: Optional[str] = None) -> bool:
        return self.footage.available(root)

    def to_dict(self) -> dict:
        out: dict = {
            "id": self.id,
            "category": self.category,
            "title": self.title,
            "footage": self.footage.to_dict(),
            "zones": list(self.zones),
            "options": dict(self.options),
            "expectations": [e.to_dict() for e in self.expectations],
            "futureImprovements": list(self.future_improvements),
        }
        for key, value in (
            ("description", self.description),
            ("profile", self.profile),
            ("notes", self.notes),
            ("addedAt", self.added_at),
        ):
            if value:
                out[key] = value
        return out


def case_from_dict(payload: dict, *, manifest_path: Optional[str] = None) -> DatasetCase:
    """Parse a `DatasetCase` manifest. Strict: an unparseable case raises rather than being skipped."""
    try:
        footage_raw = payload["footage"]
        footage = FootageReference(
            path=str(footage_raw["path"]),
            origin=str(footage_raw["origin"]),
            licence=str(footage_raw["licence"]),
            sha256=footage_raw.get("sha256"),
            duration_seconds=_opt_float(footage_raw.get("durationSeconds")),
            fps=_opt_float(footage_raw.get("fps")),
            resolution=footage_raw.get("resolution"),
            anonymisation=footage_raw.get("anonymisation"),
        )
        expectations = [
            Expectation(
                kind=str(e["kind"]),
                type=str(e["type"]),
                count=_opt_int(e.get("count")),
                count_tolerance=int(e.get("countTolerance") or 0),
                from_seconds=_opt_float(e.get("fromSeconds")),
                to_seconds=_opt_float(e.get("toSeconds")),
                zone_id=e.get("zoneId"),
                min_confidence=_opt_float(e.get("minConfidence")),
                absent=bool(e.get("absent", False)),
                notes=e.get("notes"),
            )
            for e in payload.get("expectations") or []
        ]
        return DatasetCase(
            id=str(payload["id"]),
            category=str(payload["category"]),
            title=str(payload["title"]),
            footage=footage,
            description=payload.get("description"),
            zones=list(payload.get("zones") or []),
            profile=payload.get("profile"),
            options=dict(payload.get("options") or {}),
            expectations=expectations,
            notes=payload.get("notes"),
            future_improvements=list(payload.get("futureImprovements") or []),
            added_at=payload.get("addedAt"),
            manifest_path=manifest_path,
        )
    except KeyError as exc:
        where = manifest_path or payload.get("id", "<unknown>")
        raise DatasetError(f"dataset case {where} is missing required field {exc}") from exc


class DatasetLibrary:
    """The catalogue. Loads every `<root>/<category>/cases/*.json` manifest.

    The library is **inventory, not execution**: it knows what cases exist, which have footage on this
    machine, and how the corpus is distributed across scenarios. Running a case is `evaluation.py`'s
    job — keeping those apart is what lets `vip dataset list` work in CI, where no footage is pulled.
    """

    def __init__(self, root: Optional[str] = None) -> None:
        self.root = os.path.abspath(root or DATASETS_ROOT)
        self._cases: Dict[str, DatasetCase] = {}

    # --- loading ----------------------------------------------------------

    def load(self) -> "DatasetLibrary":
        self._cases.clear()
        for category in SCENARIO_CATEGORIES:
            case_dir = os.path.join(self.root, category, "cases")
            if not os.path.isdir(case_dir):
                continue
            for name in sorted(os.listdir(case_dir)):
                if not name.endswith(".json"):
                    continue
                path = os.path.join(case_dir, name)
                with open(path, encoding="utf-8") as fh:
                    payload = json.load(fh)
                case = case_from_dict(payload, manifest_path=path)
                if case.category != category:
                    raise DatasetError(
                        f"case '{case.id}' declares category '{case.category}' but lives under "
                        f"'{category}/' — the directory is the source of truth"
                    )
                if case.id in self._cases:
                    raise DatasetError(f"duplicate dataset case id '{case.id}'")
                self._cases[case.id] = case
        return self

    def add(self, case: DatasetCase) -> DatasetCase:
        """Register a case directly. Used by tests and by in-memory corpora."""
        if case.id in self._cases:
            raise DatasetError(f"duplicate dataset case id '{case.id}'")
        self._cases[case.id] = case
        return case

    # --- inventory --------------------------------------------------------

    def __len__(self) -> int:
        return len(self._cases)

    def __iter__(self) -> Iterator[DatasetCase]:
        return iter(self.cases())

    def cases(self, *, category: Optional[str] = None, available_only: bool = False) -> List[DatasetCase]:
        out = [c for c in self._cases.values() if category is None or c.category == category]
        if available_only:
            out = [c for c in out if c.available(os.path.dirname(self.root))]
        return sorted(out, key=lambda c: (SCENARIO_CATEGORIES.index(c.category), c.id))

    def get(self, case_id: str) -> Optional[DatasetCase]:
        return self._cases.get(case_id)

    def categories(self) -> List[str]:
        return [c for c in SCENARIO_CATEGORIES if any(x.category == c for x in self._cases.values())]

    def coverage(self) -> List[dict]:
        """Per-scenario coverage — how many cases exist and how many have footage on this machine.

        This is the milestone's honest scoreboard: it makes the *gaps* visible. A scenario with zero
        cases is not a scenario the platform handles, and printing that next to one with fifteen is
        the difference between a roadmap and a claim.
        """
        base = os.path.dirname(self.root)
        out = []
        for category in SCENARIO_CATEGORIES:
            cases = [c for c in self._cases.values() if c.category == category]
            out.append(
                {
                    "category": category,
                    "cases": len(cases),
                    "withFootage": sum(1 for c in cases if c.available(base)),
                    "expectations": sum(len(c.expectations) for c in cases),
                    "negativeExpectations": sum(len(c.negative_expectations) for c in cases),
                }
            )
        return out

    def missing_footage(self) -> List[DatasetCase]:
        base = os.path.dirname(self.root)
        return [c for c in self.cases() if not c.available(base)]


def render_coverage(coverage: Sequence[dict]) -> str:
    """A one-screen coverage table. Scenarios with no cases are printed too — the empty rows are the
    most informative part of the table."""
    lines = [
        f"{'scenario':<22} {'cases':>6} {'footage':>8} {'expects':>8} {'negative':>9}",
        "-" * 57,
    ]
    for row in coverage:
        lines.append(
            f"{row['category']:<22} {row['cases']:>6} {row['withFootage']:>8} "
            f"{row['expectations']:>8} {row['negativeExpectations']:>9}"
        )
    totals = {
        key: sum(int(r[key]) for r in coverage)
        for key in ("cases", "withFootage", "expectations", "negativeExpectations")
    }
    lines.append("-" * 57)
    lines.append(
        f"{'total':<22} {totals['cases']:>6} {totals['withFootage']:>8} "
        f"{totals['expectations']:>8} {totals['negativeExpectations']:>9}"
    )
    return "\n".join(lines)


def _opt_int(value) -> Optional[int]:  # noqa: ANN001
    return None if value is None else int(value)


def _opt_float(value) -> Optional[float]:  # noqa: ANN001
    return None if value is None else float(value)
